Convert cts-cli to python

clumens opened the pull request:
This is not ready for full review.  At the least, I still need to add all the crm_shadow tests.  I am mostly pushing this now because it's a ton of work and I want it to exist in more places than just my computer.

Still, it might be worth looking around and seeing if I am on the right track here.

Edit: I meant to include this in the initial comment.  Here's a list of significant user-facing changes from the shell implementation:

* -t is now -r for consistency with other programs
* on failure to validate, add xmllint stdout/stderr to detailed output
* the Running: headers have been removed, and the Running: line for each test has been
  changed to include the regression test, in order to make the output make sense when
  multiprocessing is in use
* added an additional resource for simpler ACL testing
* commands changed:
    * `crm_mon -1 --output-as=xml` -> `crm_mon --output-as=xml`
    * `cibadmin -Ql` -> `cibadmin -Q`
    * `crm_error -- 123` -> `crm_error 123`
* some regression output has changed:
    * Passed:/Failed: lines have been changed to fix alignment problems
    * "Created new pacemaker configuration" and "A new shadow instance..." messages have been removed
    * access_render expected output now has "[0m" on every line to reset colors to normal - this causes the
      results to be in color when viewed with less/vim
    * `log_xmllib_err` -> `pcmk__log_xmllib_err`
    * standardized how XML output format is mentioned in test description
* rules tests no longer use a shadow CIB file (they didn't need to in the first place)
* tools regression test changes:
    * moved option testing code to the beginning
    * removed certain query commands, since any call with update_cib=True will also cause the CIB to
      be written to the output, so these are redundant

kgaillot reviewed (commented):
Only had time to look at the first few commits so far

clumens commented on cts/cts-cli.in:
```diff
@@ -0,0 +1,102 @@
+            if os.path.exists("%s/daemons/%s/pacemaker-%s" % (srcdir, daemon, daemon)):
+                os.environ["PATH"] = "%s:%s/daemons/%s" % (os.environ["PATH"], srcdir, daemon)
+
+        if os.path.exists("%s/xml" % srcdir):
```
There's likely to be a fair amount of this kind of stuff since I just converted big chunks from shell to python.

kgaillot replied:
We definitely don't have to clean up everything, but in this case the os.environ call below was originally under the `if`, and I'd rather drop the `if` than put it back

clumens replied:
I'm fine with cleaning up as much as we can.  There's likely to be more stuff lurking that doesn't make much sense, and I'm happy to see it go.

kgaillot reviewed (commented):
Reviewed through iso8601 commit, I'll let you muck with it before picking it up again

This is beautiful, I really like it

clumens commented on rpm/pacemaker.spec.in:
```diff
@@ -593,7 +593,7 @@ popd
 make %{_smp_mflags} check
 { cts/cts-scheduler --run load-stopped-loop \
-  && cts/cts-cli \
+  && LD_LIBRARY_PATH=lib/common/.libs PYTHONPATH=python cts/cts-cli \
```
It's the `pacemaker/python` directory containing the pacemaker module.

clumens commented on cts/cts-cli.in:
```diff
@@ -361,6 +361,89 @@ class ValidatingTest(Test):
 
 
+class TestGroup:
+    """A base class for a group of related tests."""
```
This is not directly in response to your comment, but more about TestGroups in general...

Once you get in a little farther, you're going to see that TestGroup kind of goes off the rails.  My original idea for it was to group together tests that have some common setup/teardown code, or that require the same environment variable settings.  For instance, a block of tests that all want `CIB_file=crm_mon.xml` would make a nice group.  However as things get farther along and tests get more complicated, I introduced more and more groups to do environment variable setting and overriding.

Now I'm thinking that maybe `Test.__init__` should also take an env parameter, and that anything set in that overrides anything set by the `TestGroup`, which in turn overrides whatever just happened to be in the environment in the first place.  That might clean up the code a bit.  I was initially hesitant to add environment variable stuff to individual tests, but I think now that it's probably for the best.

clumens commented on cts/cts-cli.in:
```diff
@@ -686,6 +696,215 @@ class DaemonsRegressionTest(RegressionTest):
+        """A list of Test instances to be run as part of this regression test."""
+        basic_tests = [
+            Test("Basic output", "crm_mon -1"),
+            ValidatingTest("Basic output", "crm_mon --output-as=xml"),
```
Yes and no.  The crm_mon tests require sorting out which things get included/excluded per format, and it was a bit more work than I wanted to do at the moment.  I need to get back to it.

Incidentally, I kind of hate the make_test_group thing, but I couldn't think of a better way to generate multiple tests that would only ever differ by a single command line option.

clumens commented on python/pacemaker/exitstatus.py:
```diff
@@ -60,3 +75,9 @@ class ExitStatus(IntEnum):
+def exit_str(status):
+    """Given an ExitStatus, return the matching error string."""
+    _libcrmcommon.crm_exit_str.restype = ctypes.c_char_p
+    return _libcrmcommon.crm_exit_str(status).decode()
```
It's especially worth thinking about this patch for a couple reasons:

1.  It's going to be public API, so we had better be happy with the name.
2. We're likely to use this same method in the future for adding additional bindings.

clumens replied:
Yeah, definitely.

Adding this patch on top of the original one (and I'd squash these together):
```
--- a/python/pacemaker/exitstatus.py
+++ b/python/pacemaker/exitstatus.py
@@ -1,6 +1,6 @@
 """A module providing constants relating to why a process or function exited."""
 
-__all__ = ["ExitStatus", "exit_str"]
+__all__ = ["ExitStatus"]
 __copyright__ = "Copyright 2023-2024 the Pacemaker project contributors"
 __license__ = "GNU Lesser General Public License version 2.1 or later (LGPLv2.1+)"
 
@@ -20,6 +20,7 @@ def _load_library(basename):
 
 
 _libcrmcommon = _load_library("crmcommon")
+_libcrmcommon.crm_exit_str.restype = ctypes.c_char_p
 
 
 # These values must be kept in sync with include/crm/common/results.h
@@ -76,8 +77,6 @@ class ExitStatus(IntEnum):
     NONE = 193
     MAX = 255
 
-
-def exit_str(status):
-    """Given an ExitStatus, return the matching error string."""
-    _libcrmcommon.crm_exit_str.restype = ctypes.c_char_p
-    return _libcrmcommon.crm_exit_str(status).decode()
+    def __str__(self):
+        """Given an ExitStatus, return the matching error string."""
+        return _libcrmcommon.crm_exit_str(self.value).decode()
```

Lets us do:
```
$ LD_LIBRARY_PATH=lib/common/.libs PYTHONPATH=python python
Python 3.11.6 (main, Oct  3 2023, 00:00:00) [GCC 13.2.1 20230728 (Red Hat 13.2.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from pacemaker.exitstatus import *
>>> str(ExitStatus.INVALID_PARAM)
'Invalid parameter'
>>> str(ExitStatus.SOFTWARE)
'Internal software bug'
>>> 
```

kgaillot replied:
nice

clumens commented:
Added the rest of the tests, and addressed various review comments.  I'm going to remove the WIP/Draft marks from this because I think we're ready to talk about it for real.  I will likely have more tweaks as I look at it some more.

clumens commented:
Retest this please

clumens commented:
Updated to make it easier to run things without having to add LD_LIBRARY_PATH and PYTHONPATH all over the place, as well as to remove the old regression.tools.exp file that I had overlooked.

kgaillot commented on cts/cts-cli.in:
```diff
@@ -352,6 +360,31 @@ class Test:
+
+    def _validate_hook(self, rc, stdout, stderr, valgrind=False):
+        """Validate test output with xmllint."""
+        # If the test failed, don't bother validating its results.
```
failures should still validate against the api-result schema. we might want to return the original error rather than the validation error though (or add something to the message indicating validation also failed)

clumens replied:
Fixed.

kgaillot reviewed (commented):
reviewed through make_test_group

clumens commented on cts/cts-cli.in:
```diff
@@ -22,6 +22,15 @@ other_tests = ["agents"]
+def apply_substitutions(s):
+    """Apply text substitutions to an input string and return it."""
+    substitutions = {
+        "test_home": test_home,
```
These are python format string substitutions, so they require braces, like so:

```
"cibadmin --replace --xml-file {cts_cli_data}/crm_mon-feature_set.xml",
```

kgaillot replied:
ah right, of course

clumens commented on cts/cts-cli.in:
```diff
@@ -152,15 +448,133 @@ def setup_path(opts_path):
+        if obj.name in opts.run_only:
+            regs.append(obj)
+
+    regs = run_regression_tests(regs, opts.jobs, valgrind=opts.valgrind)
```
Hah, no:

```
ValueError: Number of processes must be at least 1
```

Added a check to ensure that jobs is always at least 1.

clumens commented on python/pacemaker/exitstatus.py:
```diff
@@ -4,9 +4,25 @@
+    path = find_library(basename)
+
+    if path is None:
+        return None
```
Fixed.  Also, this code has moved to another file.  That will be in the next push here in a little bit.

clumens commented on python/pacemaker/_cts/validate.py:
```diff
@@ -0,0 +1,61 @@
+    if "PCMK_schema_directory" in os.environ:
+        return os.environ["PCMK_schema_directory"]
+
+    if os.path.exists("%s/cts-fencing.in" % sys.path[0]):
```
This is the python module search path, not $PATH (which I had to double check as well), and we can ensure it will be first based on the sys.path.insert calls at the top of cts-cli.in.  However, I can change it to search through the path instead of grabbing the first thing.

Also, this is a weird check in the first place.  I'm not sure where I came up with it.  We should probably either search some other environment variable or look for a different file.

kgaillot replied:
I think the check was copy-and-paste from elsewhere. It's fine though, all that matters is we're picking a file that's unique to pacemaker and only in a checkout (.in).

clumens commented on python/pacemaker/exitstatus.py:
```diff
@@ -4,9 +4,25 @@
+
+
+_libcrmcommon = _load_library("crmcommon")
+_libcrmcommon.crm_exit_str.restype = ctypes.c_char_p
```
I've tried to see if there's any best practices python recommends for this, without much luck.  If you grep around in /usr/lib*/python*/, you'll find that modules do all sorts of things.  Some just have a big list of these sorts of lines in a file somewhere, some do it in as part of initializing an object.  So I think we could basically do whatever we feel is best.

clumens replied:
This might be a good direction:
```
diff --git a/python/pacemaker/_library.py b/python/pacemaker/_library.py
index b9e492341f..cd6bc4ef44 100644
--- a/python/pacemaker/_library.py
+++ b/python/pacemaker/_library.py
@@ -1,6 +1,6 @@
 """A module providing private library management code."""
 
-__all__ = ["load_library"]
+__all__ = ["_libcrmcommon"]
 __copyright__ = "Copyright 2024 the Pacemaker project contributors"
 __license__ = "GNU Lesser General Public License version 2.1 or later (LGPLv2.1+)"
 
@@ -32,3 +32,7 @@ def load_library(basename):
         raise FileNotFoundError(basename)
 
     return ctypes.cdll.LoadLibrary(path)
+
+
+_libcrmcommon = load_library("crmcommon")
+_libcrmcommon.crm_exit_str.restype = ctypes.c_char_p
diff --git a/python/pacemaker/exitstatus.py b/python/pacemaker/exitstatus.py
index c699e2599e..03f7d2c8e2 100644
--- a/python/pacemaker/exitstatus.py
+++ b/python/pacemaker/exitstatus.py
@@ -4,14 +4,9 @@ __all__ = ["ExitStatus"]
 __copyright__ = "Copyright 2023-2024 the Pacemaker project contributors"
 __license__ = "GNU Lesser General Public License version 2.1 or later (LGPLv2.1+)"
 
-import ctypes
 from enum import IntEnum, unique
 
-from pacemaker._library import load_library
-
-
-_libcrmcommon = load_library("crmcommon")
-_libcrmcommon.crm_exit_str.restype = ctypes.c_char_p
+from pacemaker._library import _libcrmcommon
 
 
 # These values must be kept in sync with include/crm/common/results.h
```

Putting this stuff into `__init__.py` leads to circular import weirdness.

kgaillot replied:
looks reasonable

clumens commented on rpm/pacemaker.spec.in:
```diff
@@ -335,6 +335,7 @@ BuildRequires: %{pkgname_gettext} >= 0.18
 
 # Required for "make check"
 BuildRequires: libcmocka-devel >= 1.1.0
+BuildRequires: %{python_name}-psutil
```
I think we also need python3-psutil installed on the CI builders before this will pass.

kgaillot replied:
https://github.com/kronosnet/ci-tools/pull/27

kgaillot replied:
spelling in this commit message (can use Build: rpm:)

nrwahl2 commented:
retest this please

kgaillot commented on cts/cts-cli.in:
```diff
@@ -152,15 +456,133 @@ def setup_path(opts_path):
+    # failed output files in place, and exit with an error.  Otherwise, clean up
+    # anything that passed.
+    if test_failures > 0 and output_differs > 0:
+        print("%d test failed; see output in:" % test_failures)
```
tests (or even better check singular vs plural for here and the rest of the function)

clumens replied:
Added a solution that's potentially overkill, but potentially more generically useful.

kgaillot commented on python/pacemaker/_cts/validate.py:
```diff
@@ -0,0 +1,62 @@
+    """
+    if os.access(BuildOptions.XMLLINT_PATH, os.X_OK):
+        if rng_file is None:
+            return ["xmllint", "-"]
```
should this be XMLLINT_PATH?

kgaillot commented on python/pacemaker/_cts/validate.py:
```diff
@@ -0,0 +1,62 @@
+
+    for path in sys.path:
+        if os.path.exists("%s/cts-fencing.in" % path):
+            return "xml"
```
how does this work? wouldn't xml always be relative to the cwd?

clumens replied:
Yeah, I think the only way this makes sense is if it's `return "%s/xml" % path`.  Fixed.

kgaillot reviewed (commented):
got as far as the agent regression tests this time

kgaillot reviewed (commented):
At least skimmed the rest of the commits

"Verbosely verify a string-supplied valid configuration" takes a really long time (it didn't before)

fabbione commented:
retest this please

clumens commented:
> "Verbosely verify a string-supplied valid configuration" takes a really long time (it didn't before)

If I comment out that test, then the non-verbose one before it takes a really long time.  So that doesn't make a ton of sense.  It appears to be connected to the size of the XML file - if I change it to a smaller one, it takes a lot less time.

clumens commented on python/pacemaker/_library.py:
```diff
@@ -0,0 +1,38 @@
+    return ctypes.cdll.LoadLibrary(path)
+
+
+_libcrmcommon = load_library("crmcommon")
```
I believe it only happens once anyway.  Noting that exitstatus.py imports _libcrmcommon:

```$ git diff
diff --git a/python/pacemaker/_library.py b/python/pacemaker/_library.py
index cd6bc4ef44..db5d04cd42 100644
--- a/python/pacemaker/_library.py
+++ b/python/pacemaker/_library.py
@@ -34,5 +34,6 @@ def load_library(basename):
     return ctypes.cdll.LoadLibrary(path)
 
 
+print("RUNNING LOAD LIBRARY")
 _libcrmcommon = load_library("crmcommon")
 _libcrmcommon.crm_exit_str.restype = ctypes.c_char_p
```
```
$ LD_LIBRARY_PATH=lib/common/.libs  PYTHONPATH=python python
Python 3.12.4 (main, Jun  7 2024, 00:00:00) [GCC 14.1.1 20240607 (Red Hat 14.1.1-5)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from pacemaker._library import *
RUNNING LOAD LIBRARY
>>> from pacemaker.exitstatus import *
>>> 
```

kgaillot replied:
wow they thought that one through :)

kgaillot commented on cts/cts-cli.in:
```diff
@@ -1,3727 +1,3409 @@
+        """A list of Test instances to be run as part of this regression test."""
+        basic_tests = [
+            Test("Default standby value", "crm_standby -N node1 -G"),
+            Test("Set standby status", "crm_standby -N node1 -v true"),
```
This one does still need update_cib :)

kgaillot commented on cts/cts-cli.in:
```diff
@@ -1,3727 +1,3409 @@
+    def format_field(self, value, format_spec):
+        """Convert a value to a formatted representation."""
+        if format_spec.startswith("plural,"):
+            _, singular, plural = format_spec.split(',')
```
would be nice if plural could default to singular + s but not imporant

clumens replied:
Easy enough to add.

kgaillot reviewed (commented):
looks great :)

let me know when you're happy with it

clumens commented on cts/cts-cli.in:
```diff
@@ -1105,6 +1105,43 @@ class ErrorCodeRegressionTest(RegressionTest):
+            Test("List cluster,remote nodes",
+                 "crmadmin -N cluster,remote | wc -l | grep 9"),
+            Test("List guest,remote nodes",
+                 "crmadmin -N guest,remote | wc -l | grep 5"),
```
Easy to add, but...
```
+=#=#=#= End test: List all nodes (XML) - Failed to validate (3) =#=#=#=
+<?xml version=""?>
+<pacemaker-result api-version="X" request="crmadmin -N --output-as=xml">
+  <nodes>
+    <node type="cluster" name="overcloud-controller-0" id="1"/>
+    <node type="cluster" name="overcloud-controller-1" id="2"/>
+    <node type="cluster" name="overcloud-controller-2" id="3"/>
+    <node type="cluster" name="overcloud-galera-0" id="4"/>
+    <node type="cluster" name="overcloud-galera-1" id="5"/>
+    <node type="cluster" name="overcloud-galera-2" id="6"/>
+    <node type="guest" name="lxc1" id="lxc1"/>
+    <node type="guest" name="lxc2" id="lxc2"/>
+    <node type="remote" name="overcloud-rabbit-0" id="overcloud-rabbit-0"/>
+    <node type="remote" name="overcloud-rabbit-1" id="overcloud-rabbit-1"/>
+    <node type="remote" name="overcloud-rabbit-2" id="overcloud-rabbit-2"/>
+  </nodes>
+  <status code="0" message="OK"/>
+</pacemaker-result>
+-:1: element pacemaker-result: Relax-NG validity error : Expecting element status, got nodes
+-:1: element pacemaker-result: Relax-NG validity error : Element pacemaker-result failed to validate content
+- fails to validate
```

clumens replied:
The "cluster" and "guest" node types are not valid, according to element-crmadmin-node in crmadmin-2.25.rng.

clumens commented:
Just for a little before/after comparison:
```
before:  cts/cts-cli  38.75s user 29.46s system 103% cpu 1:05.81 total
after:   cts/cts-cli  40.60s user 19.85s system 410% cpu 14.730 total
```

kgaillot commented on cts/cli/regression.crm_verify.exp:
```diff
@@ -0,0 +1,112 @@
+* Passed: crm_verify            - Verify a string-supplied valid configuration (XML)
+=#=#=#= Begin test: Verbosely verify a string-supplied valid configuration (XML) =#=#=#=
+<pacemaker-result api-version="X" request="crm_feature_set=&quot;3.7.1&quot; transition-key=&quot;1:-1:0:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx&quot; transition-magic=&quot;0:0;1:-1:0:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx&quot; exit-reason=&quot;&quot; call-id=&quot;1&quot; rc-code=&quot;0&quot; op-status=&quot;0&quot; interval=&quot;0&quot; last-rc-change=&quot;1613491700&quot; exec-time=&quot;0&qu ...
+  <status code="0" message="OK"/>
```
it looks good on the command line, so i'm assuming one of the sed patterns is stripping it out

clumens replied:
Oh it's probably this one:
```
         (r'request=".*(crm_[a-zA-Z0-9]*)', r'request="\1'),
```

That's worth fixing before merging.

clumens commented:
Barring any other review comments, I think this is good for now.

kgaillot commented on cts/cts-cli.in:
```diff
@@ -3282,35 +3309,39 @@ def summary(regs, output_differs, verbose):
 
     if test_failures > 0:
-        print("%d tests failed" % test_failures)
+        print(fmt.format("{0} {0:plural,tests} failed", test_failures))
```
test

kgaillot reviewed (commented):
looking good
